pkg/cpe/wfn.go

```diff
@@ -0,0 +1,209 @@
+const NumAttr = 11
+
+// NonASCII reports true if the rune is not ASCII.
+func nonASCII(r rune) bool {
+	return r >= unicode.MaxASCII
+}
+
+// Reserved reports true if the rune is in the "reserved" set for CPE
+// strings and needs quoting.
+func reserved(r rune) bool {
+	return !((r >= 0x30 && r <= 0x39) ||
+		(r >= 0x41 && r <= 0x5a) ||
+		(r >= 0x61 && r <= 0x7a) ||
+		r == '_')
+}
+
+// Validate is adapted from
+// https://nvlpubs.nist.gov/nistpubs/Legacy/IR/nistir7695.pdf section 5.3.2
+func validate(s string) error {
+	if !utf8.ValidString(s) {
+		return fmt.Errorf("cpe: string not valid utf8")
+	}
+	if strings.IndexFunc(s, nonASCII) != -1 {
+		return fmt.Errorf("cpe: string contains non-ASCII characters")
+	}
+	if strings.IndexFunc(s, unicode.IsSpace) != -1 {
+		return fmt.Errorf("cpe: string contains space characters")
+	}
+	// Special character rules:
+	// A single * means ANY and a quoted hyphen would get unquoted, which means
+	// NA, so those are disallowed as string values.
+	if s == "*" {
+		return fmt.Errorf("cpe: single asterisk MUST NOT be used by itself")
+	}
+	if s == "\\-" {
+		return fmt.Errorf("cpe: quoted hyphen MUST NOT be used by itself")
+	}
+	var (
+		// In 'escaped' state
+		esc = false
+		// position of last rune
+		last = len(s) - 1
+		// if in a run of ? and if we've seen a non-? character
+		qRun, atStart = false, true
+	)
+	for i, r := range s {
+		switch r {
+		case '*':
+			if esc { // if escaped, it's not special
+				break
+			}
+			if i != 0 && i != last {
+				return fmt.Errorf("cpe: invalid position for special character: %q at %d", r, i)
+			}
+		case '?':
+			if esc { // if escaped, it's not special
+				break
+			}
+			qRun = true
+		case '\\':
+			esc = true
+			// skip resetting the esc bool
+			continue
+		default:
+			// The only place valid for a run of '?' is at the beginning and end
+			// of a string. So if we read a series of '?' then a non-'?'
+			// *twice*, that's invalid.
+			if qRun && !atStart {
+				return fmt.Errorf("cpe: invalid position for special character: %q at %d", '?', i-1)
+			}
+			if reserved(r) && !esc {
+				return fmt.Errorf("invalid unquoted character: %q at %d", r, i)
+			}
+		}
+		if r != '?' {
+			qRun, atStart = false, false
+		}
+		esc = false
+	}
+	return nil
+}
+
+// NewValue constructs a specific value and ensures it's a valid string.
+//
+// This function does not quote the provided string, only validates that the
+// quoting is proper.
+func NewValue(v string) (Value, error) {
+	if err := validate(v); err != nil {
+		return Value{}, err
+	}
+	return Value{
+		Kind: ValueSet,
+		V:    v,
+	}, nil
+}
+
+// Value represents all the states for an attribute's value.
+type Value struct {
+	Kind ValueKind
+	V    string
+}
+
+// ValueKind indicates what "kind" a value is.
+type ValueKind uint
+
+//go:generate stringer -type ValueKind
+
+// These are the valid states for a wfn attribute's value.
+const (
+	ValueUnset ValueKind = iota
+	ValueAny
+	ValueNA
+	ValueSet
```

Comment:
`ValueSet` makes me think of set-theory 'set' and may get more confusing if we plan on using adding the matching specifications into things as well. if this name does not come directly from the spec consider another term.

Reply:
I had this called `Specific` at one point, but that's also wrong because ti may be a pattern. It's a clunky name but I think should be fine along with the type.